### PR TITLE
feat: allow setting username/description during rotation

### LINKED ISSUE #69: Allow for "username" parameter on /config/rotate to change token username
We have just changed all of our admin tokens from the username "`admin`" to "`${VAULT_NAMESPACE}-vault-admin`," and had to do it all "by hand" (as in generate the new admin token and `vault write artifactory/config/admin access_token=$NEW_TOKEN`. It might be nice if people could specify a username parameter to the /config/rotate endpoint, allowing them to set/change the username for the admin token.

### PR DESCRIPTION
**Pull request by TJM**
This simply allows the optional setting of a username and/or description when rotating the admin token.

It still needs docs and tests.

Basic functionality:

```console
[tmcneely@local artifactory-secrets-plugin]$ vault read artifactory/config/admin
Key                    Value
---                    -----
access_token_sha256    d3edeabe961e3741ae0e1d8795f56e8e6fa1e562e257673c4367e58ef49f9c48
scope                  applied-permissions/admin
token_id               393984a5-2948-4fa6-bd91-49301decdbe4
url                    http://localhost:8082
use_expiring_tokens    false
username               vault-admin
version                7.55.10
[tmcneely@local artifactory-secrets-plugin]$ vault write artifactory/config/rotate username=tommy-dev-vault-admin description="Admin account for vault secrets engine"
Success! Data written to: artifactory/config/rotate
[tmcneely@local artifactory-secrets-plugin]$ vault read artifactory/config/admin
Key                    Value
---                    -----
access_token_sha256    f9c63ee191570be71fabd19ff0b3104e33ad4bc9b6970b4bbdfa05dca21c36b8
scope                  applied-permissions/admin
token_id               a94b7ea8-7588-4810-8853-80809642e002
url                    http://localhost:8082
use_expiring_tokens    false
username               tommy-dev-vault-admin
version                7.55.10
[tmcneely@local artifactory-secrets-plugin]$ curl -H "Authorization: Bearer $ACCESS_TOKEN"  http://localhost:8082/access/api/v1/tokens/a94b7ea8-7588-4810-8853-80809642e002
{
  "token_id" : "a94b7ea8-7588-4810-8853-80809642e002",
  "subject" : "jfac@01gykd5tha8tmk06x1sjmg1seb/users/tommy-dev-vault-admin",
  "issued_at" : 1682391447,
  "issuer" : "jfac@01gykd5tha8tmk06x1sjmg1seb",
  "description" : "Admin account for vault secrets engine",
  "refreshable" : false
}%
```

NOTE: The token description is stored in the database, but is not in the actual JWT token.

/closes #69

### CODE REVIEW

File: path_config_rotate.go [hunk 1]
```diff
@@ -55,6 +73,11 @@ func (b *backend) pathConfigRotateWrite(ctx context.Context, req *logical.Reques
 		Scope:    token.Scope,
 	}
 
+	// Check for new description
+	if val, ok := data.GetOk("description"); ok {
+		role.Description = val.(string)
+	}
```
Comment (TJM) [inline path_config_rotate.go:79]: Should we set a default description?

File: path_config_rotate_test.go [hunk 2]
```diff
@@ -16,3 +16,16 @@ func TestAcceptanceBackend_PathRotate(t *testing.T) {
 
 	accTestEnv.RotatePathConfig(t)
 }
+
+func TestAcceptanceBackend_PathRotateWithDetails(t *testing.T) {
```
Comment (TJM) [inline path_config_rotate_test.go:20]: I think that putting every test in the test_utils.go is going to be unmaintainable. Perhaps we should put the actual tests in the path_config_rotate_test.go, and just use the utils for the `newAcceptanceTestEnv()` and maybe even `newConfiguredAcceptanceTestEnv()` (since the vast majority of the tests will build from that)

File: path_config_rotate.go (hunk 1 above)
Comment (alexhung) [reply path_config_rotate.go:79]: Oh, that's a good idea. Something along the line of "Created by Artifactory Vault plugin", blah, blah, blah.

File: path_config_rotate_test.go (hunk 2 above)
Comment (alexhung) [reply path_config_rotate_test.go:20]: You mean func like this? https://github.com/jfrog/artifactory-secrets-plugin/blob/ed80911ba15e6f8d5db9e125a485e93960706b17/test_utils.go#L117

I'm torn TBH. On one hand, those funcs feel like they belong in that file (or rather, belong wherever `accTestEnv` struct lives, so they are in one/same place.

On the other hand, I get what you are saying as most of those funcs are used by other tests as setup steps.

File: path_config_rotate_test.go [hunk 3]
```diff
@@ -16,3 +16,16 @@ func TestAcceptanceBackend_PathRotate(t *testing.T) {
...
+	if !runAcceptanceTests {
+		t.SkipNow()
+	}
+
+	accTestEnv, err := newAcceptanceTestEnv()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	accTestEnv.RotatePathConfigWithDetails(t)
```
Comment (alexhung) [inline path_config_rotate_test.go:30]: I built the acceptance tests structure based on HashiCorp example, which is fine as a starter but probably a little simplistic. Some of the other acceptance tests have multiple func calls to setup, test, then cleanup. Ideal, I believe, is to make these funcs composable so you can mix and match as necessary.

File: test_utils.go [hunk 4]
```diff
@@ -181,6 +181,81 @@ func (e *accTestEnv) RotatePathConfig(t *testing.T) {
...
+		Data: map[string]interface{}{
+			"access_token": accessToken,
+			"url":          e.URL,
+		},
+	})
+
+	assert.NoError(t, err)
+	assert.Nil(t, resp)
+
+	// read back the path configuration and save the access token hash for comparison later
```
Comment (alexhung) [inline test_utils.go:202]: Re composability of tests, this https://github.com/jfrog/artifactory-secrets-plugin/blob/ed80911ba15e6f8d5db9e125a485e93960706b17/test_utils.go#L94 is what should be called in `TestAcceptanceBackend_PathRotateWithDetails` instead of part of this func.

File: path_config_rotate.go (hunk 1 above)
Comment (TJM) [reply path_config_rotate.go:78]: Something like that, yeah... maybe:

```
Rotated admin token, created by artifactory-secrets in vault
```

I will have to do some research to see if there is some way to obtain information about the plugin mount path, namespace, etc.

File: path_config_rotate_test.go (hunk 3 above)
Comment (alexhung) [reply path_config_rotate_test.go]: e.g.
```go
t.Run("update", accTestEnv.UpdatePathConfig)
t.Run("read", accTestEnv.ReadPathConfig)
t.Run("rotate with details", RotatePathConfigWithDetails) // New func
t.Run("read with details", ReadPathConfigWithDetails) // New func
```

File: path_config_rotate_test.go (hunk 2 above)
Comment (TJM) [reply path_config_rotate_test.go:23]: I made an update to the "config/rotate" test with what I was thinking. Basically, there is a *lot* of repeated code, so I was thinking make some utility functions, starting with `NewConfiguredAcceptanceTestEnv(t)`, but also the "read/update/delete" of the endpoints. what do you think?

Comment (alexhung) [reply path_config_rotate_test.go:23]: Looking good to me. I like them.

Comment (TJM) [reply path_config_rotate_test.go:23]: I left the old "*PathConfig" functions in as wrappers for now, but ideally those could be moved back to the path_config_test.go.... and realistically, as you say, the basic /config/admin stuff will be used in almost *every* test, so I am not sure we need to test it specifically, unless we want to test error conditions (to catch those if err != nil cases)

File: path_config_rotate.go (hunk 1 above)
Comment (TJM) [reply path_config_rotate.go]: I will leave it as a string for now, and leave the "useful details" to figure out for #43

File: test_utils.go (hunk 4 above)
Comment (TJM) [reply test_utils.go]: I was trying to make the function names more descriptive. Should I remove the wrappers and update the other calls to this?

Comment (alexhung) [reply test_utils.go]: I was referencing this pattern: https://github.com/jfrog/artifactory-secrets-plugin/pull/85#discussion_r1176770072

File: path_config_rotate_test.go (hunk 2 above)
Comment (alexhung) [reply path_config_rotate_test.go:24]: What you have is good improvement. I don't want to get into bikeshedding over minutiaes 😄

File: test_utils.go (hunk 4 above)
Comment (TJM) [reply test_utils.go]: oh I just noticed that, I suppose we can call them tests instead of just calling the functions

Comment (TJM) [reply test_utils.go]: I think maybe I am going down the wrong path...

Maybe you mean the `TestAcceptanceBackend_PathConfigRotate` test should have two *sub-tests* ... one with empty data and one with the configured username/desc? ... I can see how that might work, as they could use the same "configured test environment" ... but then it will need to be a wrapper... hmmm let me try something and see what I end up with :)

Comment (alexhung) [reply test_utils.go]: You don't have to use `t.Run()`. That will execute the func as sub-tests. In this case, you can just call the funcs and pass `t` and extra args.

Comment (TJM) [reply test_utils.go]:
```
--- PASS: TestAcceptanceBackend_PathRotate (0.12s)
    --- PASS: TestAcceptanceBackend_PathRotate/empty (0.03s)
    --- PASS: TestAcceptanceBackend_PathRotate/withDetails (0.03s)
```

File: path_config_rotate_test.go (hunk 2 above)
Comment (TJM) [reply path_config_rotate_test.go]: I think I was just mis-understanding your request.

I was too focused on using utility functions, rather than test structure :)

File: test_utils.go (hunk 4 above)
Comment (TJM) [reply test_utils.go]: ... and as far as that goes, they could have glommed onto the path_config stuff too ... I get it :)

Comment (alexhung) [reply test_utils.go]: Not exactly what I had in mind but I like it.

**Comment by TJM** [comment]
Ooops, docs (README) would be good. :)

**Review by alexhung — APPROVED** [review]
